Added option to do WTA reclustering for jet axis

jusaviin opened the pull request:
This pull request adds an option to do WTA axis reclustering for PF jets. The reclustering is disabled for calo jets. You will need to set a flag to true in the configuration file to enable the calculation of the WTA axis. I have tested this with runtest.sh script and it passes all the tests.

Commit:
Added option to do WTA reclustering for jet axis

bi-ran commented:
@jusaviin thanks!

@fhead @stepobr @cfmcginn could any one of you please check and okay this? thanks

FHead commented:
The build failed?

bi-ran commented:
@FHead no, sorry, the tests on github only works for 103X and 94X. please ignore it!

FHead commented on HeavyIonsAnalysis/JetAnalysis/interface/HiInclusiveJetAnalyzer.h:
```diff
@@ -74,6 +74,11 @@ class HiInclusiveJetAnalyzer : public edm::EDAnalyzer {
 
+  // for reWTA reclustering-----------------------
+  bool doWTARecluster_ = false;
+  fastjet::JetDefinition WTAjtDef = fastjet::JetDefinition(fastjet::JetAlgorithm::antikt_algorithm, 1, fastjet::WTA_pt_scheme);
```
Is the intention to recluster *all* constituents with WTA and get the jet axis?  If so, I would increase the 1 larger.  Imagine we run this for a wide jet and the WTA reclustering gives you two jets.

jusaviin replied:
I think that this has been planned to be used with R=0.4 jets, in which case 1 should probably be large enough radius. For wider jet radius parameters this might have to be increased.

FHead replied:
Why don't we just put 2.0 or something and don't have to worry in the future? (or even 10.0)

jusaviin replied:
That can be done This should not change the functionality for smaller jet radii?

FHead replied:
Yes, setting it arbitrary large should be fine since we collect only the candidates for the given jet.

FHead commented on HeavyIonsAnalysis/JetAnalysis/src/HiInclusiveJetAnalyzer.cc:
```diff
@@ -1464,6 +1481,25 @@ HiInclusiveJetAnalyzer::analyze(const Event& iEvent,
+
+      jets_.WTAeta[jets_.nref] = wtajt[0].eta();
+      jets_.WTAphi[jets_.nref] = wtajt[0].phi_std();
+      wtajt.clear();
```
Why do you clear this vector?  It will go out of scope in two lines...

jusaviin replied:
I think you are right and this line can be safely removed.

FHead replied:
(There is another one down below that I didn't comment on)

FHead commented on HeavyIonsAnalysis/JetAnalysis/src/HiInclusiveJetAnalyzer.cc:
```diff
@@ -247,6 +250,13 @@ HiInclusiveJetAnalyzer::beginJob() {
 
+  //for reWTA reclustering
+  // disable for calo jets
+  if(doWTARecluster_ && !(jetName_.find("Calo") != std::string::npos)){
```
Why do we disable this for calo jets?  It's an unusual use case but personally, I would rather leave the option there -- is there a forest coding guideline on these kinds of design choice?

jusaviin replied:
I needed to disable calo jets, because the test for pp and pp MC data samples failed for calo jets. The error was RefCore: A request to resolve a reference to a product of type 'reco::Candidate' with ProductID '2:55' can not be satisfied because the product cannot be found.

FHead replied:
I see - ok, in this case, let's leave it like this.  If people need WTA on calo jets for whatever reason we do another code update

FHead commented:
I added some minor comments on the code, but things look good to me generally.  I haven't try to run it myself though.

jusaviin committed:
Change WTA jet radius to 2

Based on review, the jet radius for WTA reclustering search was changed from 1 to 2

jusaviin committed:
Removed redundant clearing of wtajt vector

Cleaned the code by removing lines clearing a vector that is anyhow going out of scope.

FHead commented on HeavyIonsAnalysis/JetAnalysis/src/HiInclusiveJetAnalyzer.cc:
```diff
@@ -1464,6 +1481,25 @@ HiInclusiveJetAnalyzer::analyze(const Event& iEvent,
+      auto cs = new fastjet::ClusterSequence(candidates, WTAjtDef);
+      std::vector<fastjet::PseudoJet> wtajt = fastjet::sorted_by_pt(cs->inclusive_jets(0));
+
+      jets_.WTAeta[jets_.nref] = wtajt[0].eta();
```
I would add protection here against an empty vector.  If for some reason the list of constituents from the PF jet is empty, the code will crash since WTA jet collection is empty.

Something like
jets_.WTAeta[jets_.nref] = (wtajt.size() > 0) ? wtajt[0].eta() : -999;

jusaviin committed:
Added protection against no constituents for WTA

Added a protection for the case if for some reason no constituents for the jet can be found. In this case, instead of crashing the code, set the WTA eta and phi to -999.

FHead commented:
Things look fine from my side

bi-ran commented:
thanks @fhead @jusaviin
